pkgs/build-support/rust/default.nix

```diff
@@ -114,24 +114,48 @@ stdenv.mkDerivation (filteredArgs // {
     EOF
 
     export RUST_LOG=${logLevel}
-  '' + stdenv.lib.optionalString validateCargoDeps ''
-    if ! diff source/Cargo.lock $cargoDepsCopy/Cargo.lock ; then
-      echo
-      echo "ERROR: cargoSha256 is out of date"
-      echo
-      echo "Cargo.lock is not the same in $cargoDepsCopy"
-      echo
-      echo "To fix the issue:"
-      echo '1. Use "1111111111111111111111111111111111111111111111111111" as the cargoSha256 value'
-      echo "2. Build the derivation and wait it to fail with a hash mismatch"
-      echo "3. Copy the 'got: sha256:' value back into the cargoSha256 field"
-      echo
-
-      exit 1
-    fi
+  '' + (args.postUnpack or "");
+
+  # After unpacking and applying patches, check that the Cargo.lock matches our
+  # src package. Note that we do this after the patchPhase, because the
+  # patchPhase may create the Cargo.lock if upstream has not shipped one.
+  postPatch = (args.postPatch or "") + stdenv.lib.optionalString validateCargoDeps ''
+    (
+      cd ..
```

Comment:
You can’t assume you only need to go up one level (see crosvm for a counter-example).

But since this is now in post-patch, you’ll already be in $sourceRoot, so couldn’t you just always use Cargo.lock?

Reply:
Aha, tricky, thanks for the catch!  Indeed, you'll always be in $sourceRoot thanks to https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L1288-L1290.  That means the `Cargo.lock` file from our `src` attribute will always be at `./Cargo.lock` (I think).

The other one is the fixed-output derivation from the cargo vendor's Cargo.lock, which I *think* should always be at `$NIX_BUILD_TOP`.  Since we are using the full `$sourceRoot/Cargo.lock` env var, even if we're nested down several levels we're OK if this check just goes to the topdir.